snippets/src/lib.rs

```diff
@@ -0,0 +1,530 @@
+pub struct ByteOffset(u64);
+
+/// The number of bytes to read for the snippet from the file.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct ByteLen(u64);
+
+/// The kind of item this snippet represents.
+///
+/// # Specificity order
+///
+/// Specificity is in the order specified by the implementation of [`Ord`] for this type,
+/// meaning that a [`Kind::Full`] variant is considered a more exact match
+/// than a [`Kind::Body`] variant, which is a more exact match
+/// than a [`Kind::Signature`] variant.
+///
+/// Items with higher "specificity order" are sorted _higher_; meaning that a
+/// [`Kind::Full`] variant would be sorted later in a vector
+/// than a [`Kind::Signature`] variant:
+///
+/// ```
+/// # use snippets::*;
+/// assert!(Kind::Full > Kind::Body);
+/// assert!(Kind::Body > Kind::Signature);
+/// ```
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord, EnumIter)]
+#[non_exhaustive]
+pub enum Kind {
+    /// The signature of the function.
+    ///
+    /// ```ignore
+    /// fn say_happy_birthday(age: usize) -> String            // <- included
+    /// {                                                      // <- omitted
+    ///   println!("Happy birthday! You're {age} years old!"); // <- omitted
+    /// }                                                      // <- omitted
+    /// ```
+    Signature,
+
+    /// The body of the function.
+    ///
+    /// ```ignore
+    /// fn say_happy_birthday(age: usize) -> String {          // <- omitted
+    ///   println!("Happy birthday! You're {age} years old!"); // <- included
+    /// }                                                      // <- omitted
+    /// ```
+    Body,
+
+    /// Both signature and body.
+    ///
+    /// ```ignore
+    /// fn say_happy_birthday(age: usize) -> String {          // <- included
+    ///   println!("Happy birthday! You're {age} years old!"); // <- included
+    /// }                                                      // <- included
+    /// ```
+    Full,
+}
+
+/// The method used to generate this snippet.
+///
+/// # Specificity order
+///
+/// Specificity is in the order specified by the implementation of [`Ord`] for this type,
+/// meaning that a [`Method::Raw`] variant is considered a more exact match
+/// than a [`Method::Normalized`] variant.
+///
+/// Items with higher "specificity order" are sorted _higher_; meaning that a
+/// [`Method::Raw`] variant would be sorted later in a vector
+/// than a [`Method::Normalized`] variant:
+///
+/// ```
+/// # use snippets::*;
+/// # let arbitrary_flagset = Transforms::from(Transform::Space);
+/// assert!(Method::Raw > Method::Normalized(arbitrary_flagset));
+/// ```
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+#[non_exhaustive]
+pub enum Method {
+    /// Generated from the text with the specified normalizations applied.
+    Normalized(Transforms),
+
+    /// Generated from the text as written.
+    ///
+    /// ```ignore
```

Comment:
nit: IMO the code-block here doesn't add anything to the explanation

Reply:
Ha, good catch. That was originally meant to be compared more directly with the `Transforms` examples, but that's not really the case anymore. Will remove!

Reply:
Actually, I've made a few changes to this on my follow up branch, will just include this change there too and merge this branch as-is.